packages/maskbook/src/web3/suspends/getERC721TokenDetailed.ts

```diff
@@ -0,0 +1,44 @@
+import type { ERC721 } from '../../contracts/ERC721'
+import { formatChecksumAddress } from '../../plugins/Wallet/formatter'
+import { useERC721TokenContract } from '../contracts/useERC721TokenContract'
+import { useChainId } from '../hooks/useChainState'
+import { useSingleContractMultipleData } from '../hooks/useMulticall'
+
+import { ChainId, ERC721TokenDetailed, EthereumTokenType } from '../types'
+
+const cache = new Map<string, ERC721TokenDetailed | undefined>()
+async function getERC721TokenDetailed(
+    chainId: ChainId,
+    address: string,
+    results: any[],
+    callback: (calls_: any[]) => Promise<void>,
+    calls: any[],
+    token?: Partial<ERC721TokenDetailed>,
+) {
+    try {
+        await callback(calls)
+        // compose
+        const [name, symbol, baseURI] = results.map((x) => (x.error ? undefined : x.value))
```

Comment:
Hmm, I'm not familiar with this code flow but it seems a little strange to me. You're awaiting `callback(calls)` but don't use the return value of it, instead, you use `results` which comes from `useSingleContractMultipleData` hooks. When code recovered to run, `results` won't be updated (unless that hook is updating the original array directly). So I doubt is it working correctly?

Reply:
maybe I should cc @guanbinrui or @zhouhanseng . `results` will be updated in the second pass IIUC but this code is expected to run once for each parameter combination (then it will set the cache, in the second pass, this code pass might not be run cause it already set the cache).

Reply:
@lelenei

Reply:
When the useERC721TokenDetailed returns, the content in value is replaced

Reply:
let me check out this later

Reply:
ok I actually got some knowledge from the react team about how to write suspense code correctly. got change this later

Reply:
ok I tried to rewrite this code but found it actually delegate data fetching to `useSingleContractMultipleData`. It therefore should not be an suspense resource. It is useful when you have an async function to get the result. In this case, which is not cause you don't have an async function.

BTW I'm not meaning this is wrong:

```ts
const id = useSomeHook()
return getSuspendedData(id)
```

This is totally correct. But in this case it's not.

```ts
const [results, calls, _, callback] = useSingleContractMultipleData(erc721TokenContract, names, callData)
if (!cache.has(address)) throw getERC721TokenDetailed(chainId, address, results, callback, calls, token)
```

cc @zhouhanseng @guanbinrui @lelenei